tests/integration/test_predefined_roles.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import logging
+
+import psycopg2
+import psycopg2.sql
+import pytest
+from pytest_operator.plugin import OpsTest
+
+from .helpers import (
+    CHARM_BASE,
+    DATABASE_APP_NAME,
+    db_connect,
+    get_password,
+    get_primary,
+    get_unit_address,
+)
+from .new_relations.helpers import (
+    build_connection_string,
+    get_application_relation_data,
+    get_juju_secret,
+)
+
+logger = logging.getLogger(__name__)
+
+TIMEOUT = 15 * 60
+DATA_INTEGRATOR_APP_NAME = "data-integrator"
+
+# NOTE: We are unable to test set_user_u('operator') as dba user because psycopg2
+# runs every query in a transaction and running set_user() is not supported in transactions.
```

Comment:
I faced a similar error when trying to create a database (it cannot be done inside a transaction block). For that case, I solved by connecting to the database without a context manager (I mean, without the `with` keyword in front of either `psycopg2.connect()` or `db_connect()`, like it's done at: https://github.com/canonical/postgresql-k8s-operator/blob/fcfa631e80c4880f1e05c0a09e4f513caa1ed3e5/tests/integration/new_relations/test_new_relations_1.py#L171-L180).